### PR TITLE
four bug fixes

### PR DESCRIPTION
**Pull request by qixu001**
Fix OSS-2608 and #OSS-1579.
The root cause of OSS-2608 is that the BeginPutObject will dispose the stream but the content might be uploading in the other thread. This could happen in any thread count (not necessarily 10 thread).
The root cause of OSS-1579 is to call IDirectory.Add on an existing key when the request is retried.

Fix the OSS-2727.
Set AllowWriteStreamBuffering false when the file size is big.
Set the request.ReadWriteTimeout with the timeout value in client configuration

Fix #OSS-1393.
Support chunked encoding in PutObject.
Also automatic use chunked encoding when the stream is nonseekable.

### CODE REVIEW

File: sdk/Common/ClientConfiguration.cs [hunk 1]
```diff
@@ -28,6 +28,7 @@ public class ClientConfiguration
         private bool _isCname = false;
         private bool _enalbeMD5Check = false;
         private long _progressUpdateInterval = 1024 * 4;
+        private long _directWriteStreamThreshold = 1024 * 1024 * 512;
```
Comment (baiyubin) [inline sdk/Common/ClientConfiguration.cs]: - 默认值请修改成 8MB

File: test/TestCase/ObjectTestCase/ObjectBasicOperationTest.cs [hunk 2]
```diff
@@ -73,6 +75,99 @@ public void UploadObjectBasicSettingsTest()
...
+            finally
+            {
+                if (OssTestUtils.ObjectExists(_ossClient, _bucketName, key))
+                {
+                    _ossClient.DeleteObject(_bucketName, key);
+                }
+            }
+        }
+
+        [Test]
```
Comment (baiyubin) [inline test/TestCase/ObjectTestCase/ObjectBasicOperationTest.cs]: - 请添加一个case，GetObject的结果OssObject.Content用PutObject直接上传

File: sdk/Common/ClientConfiguration.cs (hunk 1 above)
Comment (qixu001) [reply sdk/Common/ClientConfiguration.cs]: ok

File: test/TestCase/ObjectTestCase/ObjectBasicOperationTest.cs (hunk 2 above)
Comment (qixu001) [reply test/TestCase/ObjectTestCase/ObjectBasicOperationTest.cs]: ok

**Pull request merged**